<!DOCTYPE html>
<html>
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="utf-8">
  

  
  <title>前端异常监控 | swing</title>
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  <meta name="description" content="###前言: JS异常处理简介 每天调试会看到控制台各种各式的红色错误代码。 JavaScript错误是怎么输出到控制台的  ——  JavaScript的错误处理原理。 当抛出异常时，JavaScript解释器会立即停下当前正在执行的逻辑，并且跳到最近的异常处理程序（可以参考冒泡与变量作用域的相关概念。一句话简单理解：儿子犯了错误，找老子，一层一层往上找，直到找到能够承担错误的对象为止）。如果在">
<meta name="keywords" content="技术">
<meta property="og:type" content="article">
<meta property="og:title" content="前端异常监控">
<meta property="og:url" content="http://swingboy.github.io/2018/06/15/前端异常监控/index.html">
<meta property="og:site_name" content="swing">
<meta property="og:description" content="###前言: JS异常处理简介 每天调试会看到控制台各种各式的红色错误代码。 JavaScript错误是怎么输出到控制台的  ——  JavaScript的错误处理原理。 当抛出异常时，JavaScript解释器会立即停下当前正在执行的逻辑，并且跳到最近的异常处理程序（可以参考冒泡与变量作用域的相关概念。一句话简单理解：儿子犯了错误，找老子，一层一层往上找，直到找到能够承担错误的对象为止）。如果在">
<meta property="og:locale" content="zh-CN">
<meta property="og:updated_time" content="2020-04-27T13:22:29.047Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="前端异常监控">
<meta name="twitter:description" content="###前言: JS异常处理简介 每天调试会看到控制台各种各式的红色错误代码。 JavaScript错误是怎么输出到控制台的  ——  JavaScript的错误处理原理。 当抛出异常时，JavaScript解释器会立即停下当前正在执行的逻辑，并且跳到最近的异常处理程序（可以参考冒泡与变量作用域的相关概念。一句话简单理解：儿子犯了错误，找老子，一层一层往上找，直到找到能够承担错误的对象为止）。如果在">
  
    <link rel="alternate" href="/atom.xml" title="swing" type="application/atom+xml">
  
  
    <link rel="icon" href="/favicon.png">
  
  
    <link href="//fonts.googleapis.com/css?family=Source+Code+Pro" rel="stylesheet" type="text/css">
  
  <link rel="stylesheet" href="/css/style.css">
</head>
</html>
<body>
  <div id="container">
    <div id="wrap">
      <header id="header">
  <div id="banner"></div>
  <div id="header-outer" class="outer">
    <div id="header-title" class="inner">
      <h1 id="logo-wrap">
        <a href="/" id="logo">swing</a>
      </h1>
      
    </div>
    <div id="header-inner" class="inner">
      <nav id="main-nav">
        <a id="main-nav-toggle" class="nav-icon"></a>
        
          <a class="main-nav-link" href="/">Home</a>
        
          <a class="main-nav-link" href="/archives">Archives</a>
        
      </nav>
      <nav id="sub-nav">
        
          <a id="nav-rss-link" class="nav-icon" href="/atom.xml" title="RSS Feed"></a>
        
        <a id="nav-search-btn" class="nav-icon" title="搜索"></a>
      </nav>
      <div id="search-form-wrap">
        <form action="//google.com/search" method="get" accept-charset="UTF-8" class="search-form"><input type="search" name="q" class="search-form-input" placeholder="Search"><button type="submit" class="search-form-submit">&#xF002;</button><input type="hidden" name="sitesearch" value="http://swingboy.github.io"></form>
      </div>
    </div>
  </div>
</header>
      <div class="outer">
        <section id="main"><article id="post-前端异常监控" class="article article-type-post" itemscope itemprop="blogPost">
  <div class="article-meta">
    <a href="/2018/06/15/前端异常监控/" class="article-date">
  <time datetime="2018-06-15T09:23:30.000Z" itemprop="datePublished">2018-06-15</time>
</a>
    
  </div>
  <div class="article-inner">
    
    
      <header class="article-header">
        
  
    <h1 class="article-title" itemprop="name">
      前端异常监控
    </h1>
  

      </header>
    
    <div class="article-entry" itemprop="articleBody">
      
        <p>###前言: JS异常处理简介</p>
<p>每天调试会看到控制台各种各式的红色错误代码。</p>
<p>JavaScript错误是怎么输出到控制台的  ——  JavaScript的错误处理原理。</p>
<p>当抛出异常时，JavaScript解释器会立即停下当前正在执行的逻辑，并且跳到最近的异常处理程序（可以参考冒泡与变量作用域的相关概念。一句话简单理解：儿子犯了错误，找老子，一层一层往上找，直到找到能够承担错误的对象为止）。如果在所有的词法结构和调用栈上面没有找到错误处理程序。JavaScript就会把异常当做程序错误处理，并报告给用户。</p>
<blockquote>
<p>只要有异常对象抛出，不管是浏览器抛出的，还是代码主动抛出，都会让程序停止执行。如果想让程序继续执行，则有也可以用try…catch来捕获。</p>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">&lt;script&gt;</span><br><span class="line">  error</span><br><span class="line">  console.log(&apos;永远不会执行&apos;);</span><br><span class="line">&lt;/script&gt;</span><br><span class="line">&lt;script&gt;</span><br><span class="line">  console.log(&apos;我继续执行&apos;)</span><br><span class="line">&lt;/script&gt;</span><br></pre></td></tr></table></figure>
<p>再解释一下：</p>
<blockquote>
<p>1.当前代码块将作为一个任务压入任务队列中，JavaScript线程会不断地从任务队列中提取任务执行；</p>
</blockquote>
<blockquote>
<p>2.当任务执行过程中报异常，且异常没有捕获处理，则会一路沿着调用栈从顶到底抛出，最终终止当前任务的执行；</p>
</blockquote>
<blockquote>
<p>3.JavaScript线程会继续从任务队列中提取下一个任务继续执行。</p>
</blockquote>
<h3 id="一、收集日志的方法"><a href="#一、收集日志的方法" class="headerlink" title="一、收集日志的方法"></a>一、收集日志的方法</h3><p>平时收集日志的手段，可以归类为两个方面，一个是逻辑中的错误判断，为主动判断；一个是利用语言给我们提供的捷径，暴力式获取错误信息，如 try..catch 和 window.onerror等</p>
<h5 id="1-主动判断"><a href="#1-主动判断" class="headerlink" title="1. 主动判断"></a>1. 主动判断</h5><p>我们在一些运算之后，得到一个期望的结果，然而结果不是我们想要的</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">// test.js</span><br><span class="line">function calc()&#123;</span><br><span class="line">  // sth code...</span><br><span class="line">  return val;</span><br><span class="line">&#125;</span><br><span class="line">if(calc() !== &quot;someVal&quot;)&#123;</span><br><span class="line">  Reporter.send(xxx);</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>这种属于逻辑错误/状态错误的反馈，在接口状态判断中用的比较多。</p>
<p>#####2. try..catch 捕获<br>判断一个代码段中存在的错误：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">try &#123;</span><br><span class="line">  init();</span><br><span class="line">  // code...</span><br><span class="line">&#125; catch(e)&#123;</span><br><span class="line">  Reporter.send(format(e));</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>以 init 为程序的入口，代码中所有同步执行出现的错误都会被捕获，这种方式也可以很好的避免程序刚跑起来就挂掉。</p>
<h5 id="3-window-onerror"><a href="#3-window-onerror" class="headerlink" title="3. window.onerror"></a>3. window.onerror</h5><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">window.onerror = function() &#123;</span><br><span class="line">  var errInfo = format(arguments);</span><br><span class="line">  Reporter.send(errInfo);</span><br><span class="line">  return true;</span><br><span class="line">&#125;;</span><br></pre></td></tr></table></figure>
<p>(上文中返回true时，便不会把错误暴露在控制台中)</p>
<h3 id="二、会出现的问题或注意事项或解决办法"><a href="#二、会出现的问题或注意事项或解决办法" class="headerlink" title="二、会出现的问题或注意事项或解决办法"></a>二、会出现的问题或注意事项或解决办法</h3><h5 id="关于window-onerror-的使用"><a href="#关于window-onerror-的使用" class="headerlink" title="关于window.onerror 的使用"></a>关于window.onerror 的使用</h5><figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">// test.js</span><br><span class="line">throw new Error(&quot;SHOW ME&quot;);</span><br><span class="line">window.onerror = function()&#123;</span><br><span class="line">	console.log(arguments);</span><br><span class="line">	// 阻止在控制台中打印错误信息</span><br><span class="line">	return true;</span><br><span class="line">&#125;;</span><br></pre></td></tr></table></figure>
<p>上面的代码直接报错了，没有继续往下执行。页面中可能有好几个 script 标签，但是 window.onerror 这个错误监听一定要放到最前头！</p>
<h5 id="关于-try-catch-的使用"><a href="#关于-try-catch-的使用" class="headerlink" title="关于 try..catch 的使用"></a>关于 try..catch 的使用</h5><p>对于 try..catch 的使用，一般建议是：能不用，尽量不要用。JS代码都是自己写出来的，哪里会出现问题，会出现什么问题，心中应该都有个谱，平时用到 try..catch 的一般只有两个地方：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">// JSON 格式不对</span><br><span class="line">try&#123;</span><br><span class="line">  JSON.parse(JSONString);</span><br><span class="line">&#125;catch(e)&#123;&#125;</span><br><span class="line"></span><br><span class="line">// 存在不可 decode 的字符</span><br><span class="line">try&#123;</span><br><span class="line">  decodeURIComponent(&quot;%111&quot;)</span><br><span class="line">&#125;catch(e)&#123;&#125;</span><br></pre></td></tr></table></figure>
<p>不能捕获异步的错误，promise、setTimeout等异常</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">try &#123;</span><br><span class="line">    console.log(1);</span><br><span class="line">    setTimeout(function () &#123;</span><br><span class="line">        console.log(2);</span><br><span class="line">        JSON.parse(&apos;&#123;&#123;&#125;&apos;);</span><br><span class="line">    &#125;, 1000);</span><br><span class="line">    console.log(3);</span><br><span class="line">&#125; catch (err) &#123;</span><br><span class="line">    console.log(&apos;55&apos;); //是捕获不到的</span><br><span class="line">&#125;</span><br><span class="line">console.log(4);</span><br></pre></td></tr></table></figure>
<h4 id="如何监听promise错误"><a href="#如何监听promise错误" class="headerlink" title="如何监听promise错误"></a>如何监听promise错误</h4><p>（关于事件 unhandledrejection、rejectionhandled）</p>
<blockquote>
<p>h5新增的事件；</p>
</blockquote>
<h5 id="unhandledrejection"><a href="#unhandledrejection" class="headerlink" title="unhandledrejection"></a>unhandledrejection</h5><p>当Promise 被reject并且没有得到处理的时候，会触发unhandledrejection事件。<br>unhandledrejection继承自PromiseRejectionEvent，PromiseRejectionEvent又继承自Event。因此unhandledrejection含有PromiseRejectionEvent和Event的属性和方法。</p>
<h5 id="rejectionhandled"><a href="#rejectionhandled" class="headerlink" title="rejectionhandled"></a>rejectionhandled</h5><p>当一个Promise错误最初未被处理，但是稍后又得到了处理，则会触发rejectionhandled事件。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">window.addEventListener(&apos;error&apos;, error =&gt; &#123;</span><br><span class="line">    debugger;</span><br><span class="line">    console.log(&apos;error 监听&apos;,error); // 不会触发,这里不会</span><br><span class="line">&#125;);</span><br><span class="line"></span><br><span class="line">window.addEventListener(&apos;unhandledrejection&apos;, event =&gt; &#123;</span><br><span class="line">    debugger;</span><br><span class="line">    console.log(event.reason); // 打印&quot;啦啦啦啦 reject&quot;</span><br><span class="line">&#125;);</span><br><span class="line">window.addEventListener(&apos;rejectionhandled&apos;, event =&gt; &#123;</span><br><span class="line">    debugger;</span><br><span class="line">    console.log(&apos;rejection handled&apos;);</span><br><span class="line">&#125;);</span><br><span class="line">function foo() &#123;</span><br><span class="line">    return Promise.reject(&apos;啦啦啦啦 reject&apos;);</span><br><span class="line">&#125;</span><br><span class="line">var r = foo();</span><br><span class="line"></span><br><span class="line">setTimeout(() =&gt; &#123;</span><br><span class="line">    r.catch(e =&gt; &#123; &#125;);</span><br><span class="line">&#125;, 1000);</span><br></pre></td></tr></table></figure>
<h5 id="无具体报错信息，只有Script-error"><a href="#无具体报错信息，只有Script-error" class="headerlink" title="无具体报错信息，只有Script error"></a>无具体报错信息，只有Script error</h5><blockquote>
<p>需要设置 crossorigin 并且服务器要设置 Access-Control-Allow-Origin 的响应头(被引用的资源链)：</p>
</blockquote>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">&lt;script src=&quot;http://localhost/test.js&quot; crossorigin&gt;&lt;/script&gt;</span><br></pre></td></tr></table></figure>
<blockquote>
<p>这种简单报错： Script error，目的是避免数据泄露到不安全的域中</p>
</blockquote>
<blockquote>
<p>crossOrigin参数 跳过跨域限制<br>image 和 script 标签都有此参数，它的作用就是告诉浏览器，我要加载一个外域的资源，并且我信任这个资源。</p>
</blockquote>
<h5 id="try-catch-拦截非同域的脚本错误（跨域脚本错误）"><a href="#try-catch-拦截非同域的脚本错误（跨域脚本错误）" class="headerlink" title="try catch 拦截非同域的脚本错误（跨域脚本错误）"></a>try catch 拦截非同域的脚本错误（跨域脚本错误）</h5><ul>
<li>首先：浏览器不会对try-catch起来的异常进行拦截</li>
</ul>
<p>所以，劫持原生方法：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">const nativeAddEventListener = EventTarget.prototype.addEventListener;</span><br><span class="line">EventTarget.prototype.addEventListener = function (type, func, options) &#123; // 重写原生方法</span><br><span class="line">    const wrappedFunc = function (...args) &#123; // 将回调函数包裹一层try catch</span><br><span class="line">        try &#123; </span><br><span class="line">            return func.apply(this, args);</span><br><span class="line">        &#125; catch (e) &#123;</span><br><span class="line">            const errorObj = &#123;</span><br><span class="line">                ...</span><br><span class="line">                error_name: e.name || &apos;&apos;,</span><br><span class="line">                error_msg: e.message || &apos;&apos;,</span><br><span class="line">                error_stack: e.stack || (e.error &amp;amp;&amp;amp; e.error.stack),</span><br><span class="line">                error_native: e,</span><br><span class="line">                ...</span><br><span class="line">            &#125;;</span><br><span class="line">            // 接下来可以将errorObj统一进行处理。</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    return nativeAddEventListener.call(this, type, wrappedFunc, options); // 调用原生的方法，保证addEventListener正确执行</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<h5 id="async-await-和捕获异常"><a href="#async-await-和捕获异常" class="headerlink" title="async/await 和捕获异常"></a>async/await 和捕获异常</h5><p>作为异步的中级方案，catch异常也是很简单的：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">var sleep = function (time) &#123;</span><br><span class="line">    return new Promise(function (resolve, reject) &#123;</span><br><span class="line">        setTimeout(function () &#123;</span><br><span class="line">            reject(&apos;error&apos;);</span><br><span class="line">        &#125;, time);</span><br><span class="line">    &#125;)</span><br><span class="line">&#125;;</span><br><span class="line"></span><br><span class="line">var func = async function () &#123;</span><br><span class="line">    try &#123;</span><br><span class="line">        console.log(&apos;start&apos;);</span><br><span class="line">        await sleep(1000); // 这里得到了一个返回错误</span><br><span class="line">        </span><br><span class="line">        // 所以以下代码不会被执行了</span><br><span class="line">        console.log(&apos;end&apos;);</span><br><span class="line">    &#125; catch (err) &#123;</span><br><span class="line">        console.log(&apos;catch内容：&apos;, err); // 这里捕捉到错误 `error`</span><br><span class="line">    &#125;</span><br><span class="line">&#125;;</span><br><span class="line">func();</span><br></pre></td></tr></table></figure>
<p>#####有iframe的页面怎么监听？</p>
<ul>
<li>同源</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">window.frames[0].onerror = function (msg, url, row, col, error) &#123;</span><br><span class="line">    console.log(&#123;</span><br><span class="line">      msg,  url,  row, col, error</span><br><span class="line">    &#125;)</span><br><span class="line">    return true;</span><br><span class="line">&#125;;</span><br></pre></td></tr></table></figure>
<ul>
<li>非同源（包括可控与不可控）<br>可控的类似可以实行跨域那一套。postMessage等等。</li>
</ul>
<p>略</p>
<h4 id="关于网络请求的"><a href="#关于网络请求的" class="headerlink" title="关于网络请求的"></a>关于网络请求的</h4><p>由于网络请求异常不会冒泡，因此必须在capture阶段捕获才可以。但还有一个问题是这种方式无法精确判断异常的HTTP状态是404还是500等，因此还是要配合服务端日志来排查分析才可以。</p>
<p>所以，这种情况下onerror 是无法帮助我们捕获到异常的</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">&lt;script&gt;</span><br><span class="line">  window.onerror = function (msg, url, row, col, error) &#123;</span><br><span class="line">    console.log(&apos;我知道异步错误了&apos;);</span><br><span class="line">    console.log(&#123;</span><br><span class="line">      msg,  url,  row, col, error</span><br><span class="line">    &#125;)</span><br><span class="line">    return true;</span><br><span class="line">  &#125;;</span><br><span class="line">&lt;/script&gt;</span><br><span class="line">//html</span><br><span class="line">&lt;img src=&quot;./404.png&quot;&gt;</span><br></pre></td></tr></table></figure>
<p>必须要：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">window.addEventListener(&quot;error&quot;, function(e)&#123;</span><br><span class="line">    debugger;</span><br><span class="line">    console.log(&apos;报错了啦啦&apos;, e.bubbles)</span><br><span class="line">&#125;,   true);</span><br><span class="line"></span><br><span class="line">//html</span><br><span class="line">&lt;img src=&quot;404.jpg&quot;/&gt;</span><br></pre></td></tr></table></figure>
<h4 id="框架里面的问题"><a href="#框架里面的问题" class="headerlink" title="框架里面的问题"></a>框架里面的问题</h4><p>现在越来越多的项目开始使用前端框架，在MVVM框架中如果你一如既往的想使用window.onerror来捕获异常，很可能不好使，或许根本捕获不到，因为你的异常信息被框架自身的异常机制捕获了。<br>比如Vue 中我们应该这样捕获全局异常：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">Vue.config.errorHandler = function (err, vm, info) &#123;</span><br><span class="line">    let &#123; </span><br><span class="line">        message, // 异常信息</span><br><span class="line">        name, // 异常名称</span><br><span class="line">        script,  // 异常脚本url</span><br><span class="line">        line,  // 异常行号</span><br><span class="line">        column,  // 异常列号</span><br><span class="line">        stack  // 异常堆栈信息</span><br><span class="line">    &#125; = err;</span><br><span class="line"></span><br><span class="line">    // vm为抛出异常的 Vue 实例</span><br><span class="line">    // info为 Vue 特定的错误信息，比如错误所在的生命周期钩子</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<p>还有，react 16里面提供了componentDidCatch方法，可以直接捕获react组件里面的异常来直接上报。这样其实对于react的场景就更简单了。它能捕获在子组件树中任何地方的 JavaScript 异常，并打印这些错误和展示备用UI, 就像将 children 包裹在一个大的 try/catch 语句块中</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line">let Com = function()&#123;</span><br><span class="line">    return (&lt;div&gt;        </span><br><span class="line">        &lt;ErrorHandler&gt;</span><br><span class="line">            &lt;TestState/&gt;</span><br><span class="line">        &lt;/ErrorHandler&gt;</span><br><span class="line">    &lt;/div&gt;)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">ReactDOM.render(&lt;Com/&gt;, document.getElementById(&apos;root&apos;));</span><br><span class="line"></span><br><span class="line">class ErrorHandler extends React.Component &#123;</span><br><span class="line">    constructor(props) &#123;</span><br><span class="line">        super(props);</span><br><span class="line">        this.state = &#123; hasError: false &#125;;</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    componentDidCatch(error, info) &#123;</span><br><span class="line">        console.log(&apos;err.&apos;, error);</span><br><span class="line">        this.setState(&#123; hasError: true &#125;);</span><br><span class="line">        // report(error, info);</span><br><span class="line">    &#125;</span><br><span class="line"></span><br><span class="line">    render() &#123;</span><br><span class="line">        if (this.state.hasError) &#123;</span><br><span class="line">            return &lt;h1&gt;这是报错页面&lt;/h1&gt;;</span><br><span class="line">        &#125;</span><br><span class="line">        return this.props.children;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">//在子组件中</span><br><span class="line">//throw Error(&apos;我错啦&apos;);</span><br></pre></td></tr></table></figure>
<h2 id="定位错误"><a href="#定位错误" class="headerlink" title="定位错误"></a>定位错误</h2><p>我们的业务代码经webpack、gulp等打包压缩后产生如下代码：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">!function(n)&#123;function t(e)&#123;if(r[e])return r[e].exports;var o=r[e]=&#123;i:e,l:!1,exports:&#123;&#125;&#125;;</span><br></pre></td></tr></table></figure>
<p>在控制台会看到对应的数据：<br>代码如期报错，并上报相关信息</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">&#123; msg: &apos;Uncaught ReferenceError: xxx is not defined&apos;,</span><br><span class="line">  url: &apos;http://xxx/xxx.min.js&apos;,</span><br><span class="line">  row: &apos;10&apos;,</span><br><span class="line">  col: &apos;1000&apos; &#125;</span><br></pre></td></tr></table></figure>
<p>结合压缩后的代码，肉眼观察很难定位出具体问题</p>
<p>那我们如何定位到错误呢？</p>
<p>方法一：不压缩 js 代码</p>
<p>方法二：将压缩代码中分号变成换行</p>
<p>方案三：js 代码半压缩(保留空格和换行)</p>
<p>uglifyjs 的另一配置参数 beautify 设置为 true 时，最终代码将呈现压缩后进行格式化的效果（保留空格和换行），如</p>
<p>方案四：SourceMap 快速定位</p>
<blockquote>
<p>我们目前的办法是用chrome的DevTools这样的辅助工具。而且并不是每个线上资源都会添加 sourceMap 文件</p>
</blockquote>
<h5 id="sourceMap"><a href="#sourceMap" class="headerlink" title="sourceMap"></a>sourceMap</h5><p>SourceMap 是一个信息文件，存储着源文件的信息及源文件与处理后文件的映射关系。 里面储存着位置信息。也就是说，转换后的代码的每一个位置，所对应的转换前的位置。</p>
<p>我们在定位压缩代码的报错时，可以通过错误信息的行列数与对应的 SourceMap 文件，处理后得到源文件的具体错误信息。</p>
<p>打开Source map文件，它大概是这个样子</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">&#123;</span><br><span class="line">	version : 3,</span><br><span class="line">　　file: &quot;out.js&quot;,</span><br><span class="line">　　sourceRoot : &quot;&quot;,</span><br><span class="line">　　sources: [&quot;a.js&quot;, &quot;b.js&quot;],</span><br><span class="line">　　names: [&quot;src&quot;, &quot;maps&quot;, &quot;are&quot;, &quot;fun&quot;],</span><br><span class="line">　　mappings: &quot;AAgBC,SAAQ,CAAEA&quot;</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>
<ul>
<li>version：Source map的版本，目前为3。</li>
<li>file：转换后的文件名。</li>
<li>sourceRoot：转换前的文件所在的目录。如果与转换前的文件在同一目录，该项为空。</li>
<li>sources：转换前的文件。该项是一个数组，表示可能存在多个文件合并。</li>
<li>names：转换前的所有变量名和属性名。</li>
<li>mappings：记录位置信息的字符串。</li>
</ul>
<p>应用举例：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">var fs = require(&apos;fs&apos;);</span><br><span class="line">var sourceMap = require(&apos;source-map&apos;);</span><br><span class="line"></span><br><span class="line">// map文件</span><br><span class="line">var rawSourceMapJsonData = fs.readFileSync(&apos;./index.min.js.map&apos;, &apos;utf-8&apos;)</span><br><span class="line">rawSourceMapJsonData = JSON.parse(rawSourceMapJsonData)</span><br><span class="line"></span><br><span class="line">var consumer = new sourceMap.SourceMapConsumer(rawSourceMapJsonData);</span><br><span class="line"></span><br><span class="line">// 得到真实错误位置</span><br><span class="line">console.log(consumer.originalPositionFor(&#123;line: 1, column: 100&#125;))</span><br></pre></td></tr></table></figure>
<p><a href="http://www.ruanyifeng.com/blog/2013/01/javascript_source_map.html" target="_blank" rel="noopener">阮一峰JavaScript Source Map 详解</a></p>
<p><a href="https://github.com/swingboy/errorMonitor" target="_blank" rel="noopener">示例代码</a></p>

      
    </div>
    <footer class="article-footer">
      <a data-url="http://swingboy.github.io/2018/06/15/前端异常监控/" data-id="cka7xvr5300180oo7asvoc7zv" class="article-share-link">Share</a>
      
      
  <ul class="article-tag-list"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/技术/">技术</a></li></ul>

    </footer>
  </div>
  
    
<nav id="article-nav">
  
    <a href="/2018/06/18/node require与webpack模块化原理/" id="article-nav-newer" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Newer</strong>
      <div class="article-nav-title">
        
          node require与webpack模块化原理
        
      </div>
    </a>
  
  
    <a href="/2018/04/12/关于前端路由的一些事/" id="article-nav-older" class="article-nav-link-wrap">
      <strong class="article-nav-caption">Older</strong>
      <div class="article-nav-title">关于前端路由的一些事</div>
    </a>
  
</nav>

  
</article>

</section>
        
          <aside id="sidebar">
  
    

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签</h3>
    <div class="widget">
      <ul class="tag-list"><li class="tag-list-item"><a class="tag-list-link" href="/tags/技术/">技术</a></li><li class="tag-list-item"><a class="tag-list-link" href="/tags/随笔/">随笔</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">标签云</h3>
    <div class="widget tagcloud">
      <a href="/tags/技术/" style="font-size: 20px;">技术</a> <a href="/tags/随笔/" style="font-size: 10px;">随笔</a>
    </div>
  </div>

  
    
  <div class="widget-wrap">
    <h3 class="widget-title">归档</h3>
    <div class="widget">
      <ul class="archive-list"><li class="archive-list-item"><a class="archive-list-link" href="/archives/2020/01/">一月 2020</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/12/">十二月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2019/03/">三月 2019</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/11/">十一月 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/08/">八月 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/07/">七月 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/06/">六月 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/04/">四月 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/03/">三月 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2018/02/">二月 2018</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/12/">十二月 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/10/">十月 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2017/07/">七月 2017</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/11/">十一月 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/06/">六月 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/05/">五月 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/04/">四月 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2016/02/">二月 2016</a></li><li class="archive-list-item"><a class="archive-list-link" href="/archives/2015/10/">十月 2015</a></li></ul>
    </div>
  </div>


  
    
  <div class="widget-wrap">
    <h3 class="widget-title">最新文章</h3>
    <div class="widget">
      <ul>
        
          <li>
            <a href="/2020/01/24/简单梳理react优化相关/">简单梳理react优化相关</a>
          </li>
        
          <li>
            <a href="/2019/12/10/关于前端性能优化的一些事/">关于前端性能优化的一些事</a>
          </li>
        
          <li>
            <a href="/2019/03/29/webpack原理/">webpack原理</a>
          </li>
        
          <li>
            <a href="/2019/03/29/webpackPlugin/">webpackPlugin</a>
          </li>
        
          <li>
            <a href="/2019/03/29/webpackLoader/">webpackLoader</a>
          </li>
        
      </ul>
    </div>
  </div>

  
</aside>
        
      </div>
      <footer id="footer">
  
  <div class="outer">
    <div id="footer-info" class="inner">
      &copy; 2020 火柴人<br>
      Powered by <a href="http://hexo.io/" target="_blank">Hexo</a>
    </div>
  </div>
</footer>
    </div>
    <nav id="mobile-nav">
  
    <a href="/" class="mobile-nav-link">Home</a>
  
    <a href="/archives" class="mobile-nav-link">Archives</a>
  
</nav>
    

<script src="//ajax.googleapis.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script>


  <link rel="stylesheet" href="/fancybox/jquery.fancybox.css">
  <script src="/fancybox/jquery.fancybox.pack.js"></script>


<script src="/js/script.js"></script>



  </div>
</body>
</html>